power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -58,29 +28,114 @@ std::tuple<Idx, Idx> count_voltage_sensors(const Idx n_bus, const MeasuredValues
+            } else if (flow_sensors[bus_entry_upstream] == 1) {
+                // if not possible, steal from upstream bus
+                flow_sensors[bus_entry_upstream] = 0;
+                flow_sensors[branch_entry_upstream] = 1;
+            }
+        }
+        // remove the current bus injection regardless of the original state
+        flow_sensors[bus_entry_current] = 0;
+    }
+    // set last bus injection to zero
+    flow_sensors[y_bus_structure.bus_entry[n_bus - 1]] = 0;
+}
 
-    Idx const n_bus{topo->n_bus()};
-    std::vector<BranchIdx> const& branch_bus_idx{topo->branch_bus_idx};
+} // namespace detail
+template <symmetry_tag sym>
+inline void necessary_observability_check(MeasuredValues<sym> const& measured_values, MathModelTopology const& topo,
+                                          YBusStructure const& y_bus_structure) {
+    Idx const n_bus{topo.n_bus()};
 
     auto const [n_voltage_sensor, n_voltage_phasor_sensor] = detail::count_voltage_sensors(n_bus, measured_values);
     if (n_voltage_sensor < 1) {
-        throw NotObservableError{"no voltage sensor found"};
+        throw NotObservableError{"No voltage sensor found!\n"};
     }
 
-    Idx const n_injection_sensor = detail::count_bus_injection_sensors(n_bus, measured_values);
-    Idx const n_branch_sensor = detail::count_branch_sensors(branch_bus_idx, n_bus, measured_values);
-    Idx const n_power_sensor = n_injection_sensor + n_branch_sensor;
+    std::vector<int8_t> flow_sensors = detail::count_flow_sensors(measured_values, topo, y_bus_structure);
+    // count flow sensors, note we manually specify the intial value type to avoid overflow
+    Idx const n_flow_sensor = std::reduce(flow_sensors.cbegin(), flow_sensors.cend(), Idx{}, std::plus<Idx>{});
 
-    if (n_voltage_phasor_sensor == 0 && n_power_sensor < n_bus - 1) {
+    // check nessessary condition for observability
+    if (n_voltage_phasor_sensor == 0 && n_flow_sensor < n_bus - 1) {
         throw NotObservableError{};
     }
-    if (n_voltage_phasor_sensor > 0 && n_power_sensor + n_voltage_phasor_sensor < n_bus) {
+    if (n_voltage_phasor_sensor > 0 && n_flow_sensor + n_voltage_phasor_sensor < n_bus) {
         throw NotObservableError{};
     }
+
+    // for radial grid without phasor measurement, try to assign injection sensor to branch sensor
+    // we can then check sufficient condition for observability
+    if (topo.is_radial && n_voltage_phasor_sensor == 0) {
+        detail::assign_injection_sensor(y_bus_structure, flow_sensors);
+        // count flow sensors again
+        Idx const n_flow_sensor_new = std::reduce(flow_sensors.cbegin(), flow_sensors.cend(), Idx{}, std::plus<Idx>{});
+        if (n_flow_sensor_new < n_bus - 1) {
+            throw NotObservableError{"The number of power sensors seems to be sufficient, but they are not independent "
+                                     "enough. The system is still not observable.\n"};
```

Comment:
"seems to be" is rather vague, in this case we do know the number is indeed enough. Perhaps only:
```suggestion
            throw NotObservableError{"The number of power sensors appears sufficient, but they are not independent "
                                     "enough. The system is still not observable.\n"};
```